Userland/Libraries/LibWeb/HTML/HTMLFormElement.cpp

```diff
@@ -815,4 +851,176 @@ void HTMLFormElement::plan_to_navigate_to(AK::URL url, Variant<Empty, String, PO
     VERIFY(m_planned_navigation);
 }
 
+// https://html.spec.whatwg.org/multipage/forms.html#the-form-element:supported-property-indices
+bool HTMLFormElement::is_supported_property_index(u32 index) const
+{
+    // The supported property indices at any instant are the indices supported by the object returned by the elements attribute at that instant.
+    return index < elements()->length();
+}
+
+// https://html.spec.whatwg.org/multipage/forms.html#dom-form-item
+WebIDL::ExceptionOr<JS::Value> HTMLFormElement::item_value(size_t index) const
+{
+    // To determine the value of an indexed property for a form element, the user agent must return the value returned by
+    // the item method on the elements collection, when invoked with the given index as its argument.
+    return elements()->item(index);
+}
+
+// https://html.spec.whatwg.org/multipage/forms.html#the-form-element:supported-property-names
+Vector<FlyString> HTMLFormElement::supported_property_names() const
+{
+    // The supported property names consist of the names obtained from the following algorithm, in the order obtained from this algorithm:
+
+    // 1. Let sourced names be an initially empty ordered list of tuples consisting of a string, an element, a source,
+    //    where the source is either id, name, or past, and, if the source is past, an age.
+    struct SourcedName {
+        FlyString name;
+        JS::GCPtr<DOM::Element const> element;
+        enum class Source {
+            Id,
+            Name,
+            Past,
+        } source;
+        Duration age;
+    };
+    Vector<SourcedName> sourced_names;
+
+    // 2. For each listed element candidate whose form owner is the form element, with the exception of any
+    //    input elements whose type attribute is in the Image Button state:
+    for (auto const& candidate : m_associated_elements) {
+        if (!is_form_control(*candidate, *this))
+            continue;
+
+        // 1. If candidate has an id attribute, add an entry to sourced names with that id attribute's value as the
+        //    string, candidate as the element, and id as the source.
+        if (candidate->has_attribute(HTML::AttributeNames::id))
+            sourced_names.append(SourcedName { candidate->id().value(), candidate, SourcedName::Source::Id, {} });
+
+        // 2. If candidate has a name attribute, add an entry to sourced names with that name attribute's value as the
+        //    string, candidate as the element, and name as the source.
+        if (candidate->has_attribute(HTML::AttributeNames::name))
+            sourced_names.append(SourcedName { candidate->attribute(HTML::AttributeNames::name).value(), candidate, SourcedName::Source::Name, {} });
```

Comment:
Pretty minor, but for all of these has_attribute checks we can save a second lookup in the named node map by:

```suggestion
        if (auto maybe_name = candidate->attribute(HTML::AttributeNames::name); maybe_name.has_value())
            sourced_names.append(SourcedName { candidate->maybe_name.release_value(), candidate, SourcedName::Source::Name, {} });
```

And for `has_attribute` on ID that can similarly be:
```cpp
        if (auto maybe_id = candidate->id(); maybe_id.has_value())
            sourced_names.append(SourcedName { maybe_id.release_value(), candidate, SourcedName::Source::Id, {} });
```